src/System.Text.Json/src/System/Text/Json/JsonReaderHelper.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    internal static class JsonReaderHelper
+    {
+        public static (int, int) CountNewLines(ReadOnlySpan<byte> data)
+        {
+            int lastLineFeedIndex = -1;
+            int newLines = 0;
+            for (int i = 0; i < data.Length; i++)
+            {
+                if (data[i] == JsonConstants.LineFeed)
+                {
+                    lastLineFeedIndex = i;
+                    newLines++;
+                }
+            }
+            return (newLines, lastLineFeedIndex);
+        }
+
+        // https://tools.ietf.org/html/rfc8259
+        // Does the span contain '\' or any control characters (i.e. 0 to 31)
+        // IndexOfAny(92, < 32)
+        // Borrowed and modified from SpanHelpers.Byte:
+        // https://github.com/dotnet/corefx/blob/fc169cddedb6820aaabbdb8b7bece2a3df0fd1a5/src/Common/src/CoreLib/System/SpanHelpers.Byte.cs#L473-L604
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int IndexOfAnyControlOrEscape(this ReadOnlySpan<byte> span)
+        {
+            return IndexOfAny(
+                    ref MemoryMarshal.GetReference(span),
+                    JsonConstants.ReverseSolidus,
+                    JsonConstants.Space,
+                    span.Length);
+        }
+
+        private static unsafe int IndexOfAny(ref byte searchSpace, byte value0, byte lessThan, int length)
```

Comment:
Why can't the built-in IndexOfAny be used?

Reply:
Ah, I see, it's specialized with the "lessThan".

Reply:
It's an optimization to do a quick (vectorized) to know whether it is necessary to validate the bytes representing the string JSON token for potential invalid characters. Otherwise, we can safely assume the token is valid and can skip validation (according to the JSON RFC).

The validation is a non-vectorized O(n) operation. There is potentially a better way to do this such that we avoid passing through the bytes twice and still be fast in the common, valid case.

Reply:
It might be good to add comments explaining this.

Reply:
Or give it a different name eg IndexOfOrLessThan...its a poor name but it's not as misleading as IndexOfAny whose behavior I make assumptions about.